From 5ebb33c6547491f71af1950974e058aca25fa44a Mon Sep 17 00:00:00 2001
From: Joshua Bachmeier <joshua@bachmeier.cc>
Date: Sun, 1 May 2022 00:01:14 +0200
Subject: [PATCH 2/2] Add support for oauth scopes

As defined in RFC6749:
* https://www.rfc-editor.org/rfc/rfc6749#section-4.1.1
* https://www.rfc-editor.org/rfc/rfc6749#section-3.3

Scopes can be configured per-provider in engelsystem config.
---
 config/config.default.php           |  2 ++
 src/Controllers/OAuthController.php | 17 +++++++++++++++--
 2 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/config/config.default.php b/config/config.default.php
index 8b279c65..d0c07a95 100644
--- a/config/config.default.php
+++ b/config/config.default.php
@@ -82,6 +82,8 @@ return [
             'url_token' => '[generated by provider]',
             // User info URL which provides userdata
             'url_info' => '[generated by provider]',
+            // OAuth Scopes
+            'scope' => ['openid'],
             // Info unique user id field
             'id' => 'uuid',
             // The following fields are used for registration
diff --git a/src/Controllers/OAuthController.php b/src/Controllers/OAuthController.php
index 93b6012d..ecb0021f 100644
--- a/src/Controllers/OAuthController.php
+++ b/src/Controllers/OAuthController.php
@@ -87,6 +87,7 @@ class OAuthController extends BaseController
     {
         $providerName = $request->getAttribute('provider');
         $provider = $this->getProvider($providerName);
+        $config = $this->config->get('oauth')[$providerName];
 
         // Handle OAuth error response according to https://www.rfc-editor.org/rfc/rfc6749#section-4.1.2.1
         if ($request->has('error')) {
@@ -96,7 +97,13 @@ class OAuthController extends BaseController
         }
 
         if (!$request->has('code')) {
-            $authorizationUrl = $provider->getAuthorizationUrl();
+            $authorizationUrl = $provider->getAuthorizationUrl(
+                [
+                    // Leauge seperates scopes by comma, which is wrong, so we do it
+                    // here properly by spaces. See https://www.rfc-editor.org/rfc/rfc6749#section-3.3
+                    'scope' => join(' ', $config['scope'])
+                ]
+            );
             $this->session->set('oauth2_state', $provider->getState());
 
             return $this->redirector->to($authorizationUrl);
@@ -181,7 +188,13 @@ class OAuthController extends BaseController
         }
 
         $config = $this->config->get('oauth')[$providerName];
-        $userdata = new Collection($resourceOwner->toArray());
+        $resourceData = $resourceOwner->toArray();
+        if (!empty($config['nested_info'])) {
+            $resourceData = Arr::dot($resourceData);
+        }
+
+        $userdata = new Collection($resourceData);
+
         if (!$oauth) {
             return $this->redirectRegisterOrThrowNotFound(
                 $providerName,
-- 
2.33.1

